### PR TITLE
[DeviceSanitizer] Only try to get backtrace symbols when needed

### PR DESCRIPTION
**Pull request by zhaomaosu**

### CODE REVIEW

File: source/loader/layers/sanitizer/stacktrace.hpp [hunk 1]
```diff
@@ -18,14 +18,16 @@
 
 namespace ur_sanitizer_layer {
 
-constexpr size_t MAX_BACKTRACE_FRAMES = 64;
+constexpr size_t MAX_BACKTRACE_FRAMES = 48;
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/stacktrace.hpp:21]: why this change?

File: source/loader/layers/sanitizer/linux/backtrace.cpp [hunk 2]
```diff
@@ -17,20 +17,20 @@ namespace ur_sanitizer_layer {
 StackTrace GetCurrentBacktrace() {
     void *Frames[MAX_BACKTRACE_FRAMES];
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/linux/backtrace.cpp:18]: Using `BacktraceFrame` type for `Frames`?

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/linux/backtrace.cpp]: Done

File: source/loader/layers/sanitizer/stacktrace.hpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/stacktrace.hpp]: I originally think 64 is too large, but looks like it will not impact performance. Revert this change.

**Review by yingcong-wu — APPROVED** [review]

**Review by AllanZyne — APPROVED** [review]

**Comment by zhaomaosu** [comment]
Hi @oneapi-src/unified-runtime-maintain, could you please help review this PR? Thanks.

File: source/loader/layers/sanitizer/linux/backtrace.cpp [hunk 3]
```diff
@@ -15,22 +15,22 @@
...
-        return StackTrace();
-    }
 
     StackTrace Stack;
-    for (int i = 0; i < FrameCount; i++) {
-        BacktraceInfo addr_info(Symbols[i]);
-        Stack.stack.emplace_back(addr_info);
-    }
-    free(Symbols);
+    Stack.stack =
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/linux/backtrace.cpp:22]: nit, we should be able to skip the stack buffer by writing to the vector directly:
```
StackTrace Stack;
Stack.stack.reserve(MAX_BACKTRACE_FRAMES);
int FrameCount = backtrace(Stack.stack.data(), MAX_BACKTRACE_FRAMES);
```

**Review by pbalcer — APPROVED** [review]
just one minor suggestion, feel free to ignore.

File: source/loader/layers/sanitizer/linux/backtrace.cpp (hunk 3 above)
Comment (pbalcer) [reply source/loader/layers/sanitizer/linux/backtrace.cpp:22]: I didn't want to delay merging because of this. If you want to change it, include it in another PR.

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/linux/backtrace.cpp:22]: Thanks. I'll do this later.
